Update SDK to 1.12.0-rc1

pichlermarc opened the pull request:
## Summary

This PR updates the SDK version to the latest version (1.12.0-rc1), and updates the tests accordingly.

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
+
+    def _histogram_to_dynatrace_metric(self, metric: Metric,
+                                       point: HistogramDataPoint):
+        return self._metric_factory.create_float_summary(
```
Need to check whether its cumulative or delta. can also be done in a separate PR

pichlermarc replied:
Now checking and dropping and logging cumulatives as of [2d01775](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/2d01775f272252daa002b671d1a760238bbcfc8c).

pirgeo commented on test/test_exporter.py:
```diff
@@ -158,11 +174,11 @@ def test_with_prefix(self, mock_post):
     def test_with_attributes(self, mock_post):
```
```suggestion
    def test_with_default_dimensions(self, mock_post):
```

pirgeo replied:
Do we also have a test with data attributes (where the attributes are on the data point)?

pirgeo commented on test/test_exporter.py:
```diff
@@ -204,103 +220,51 @@ def test_dynatrace_metadata_enrichment_with_default_attributes(
+        ("",),
+        (".",)
+    ])
+    def test_invalid_metricname_skipped(self, instrument_name):
```
Does this test test that one metric is dropped, but also that the chunking works? Is there any reason to have the chunking test here as well?

pirgeo commented on test/test_exporter.py:
```diff
@@ -354,11 +317,12 @@ def test_sum_delta(self, mock_post):
 
     @patch.object(requests.Session, 'post')
     def test_sum_delta_non_monotonic(self, mock_post):
```
I think this is not correct, but this also falls into the area of a potential follow up

pichlermarc replied:
Yes, this was not correct. As of [2d01775](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/2d01775f272252daa002b671d1a760238bbcfc8c), non monotonic delta sums are now dropped and test are updated accordingly. Thanks for catching this.

pirgeo commented on README.md:
```diff
@@ -1,8 +1,12 @@
+This exporter allows exporting metrics created using the [OpenTelemetry SDK for Python](https://github.com/open-telemetry/opentelemetry-python)
+directly to [Dynatrace](https://www.dynatrace.com).
+
+It was built against OpenTelemetry SDK version 1.12-rc1 and should work with any 1.12+ version.
```
```suggestion
It was built against OpenTelemetry SDK version `1.12-rc1` and should work with any `1.12+` version.
```

pichlermarc replied:
Applied in [b116247](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/b1162475784b61144a57f51243ae2df4b469104a).

pirgeo commented on README.md:
```diff
@@ -16,20 +20,20 @@ pip install opentelemetry-exporter-dynatrace-metrics
 # setup metrics export pipeline
-_metrics.set_meter_provider(MeterProvider(
+metrics.set_meter_provider(MeterProvider(
         metric_readers=[PeriodicExportingMetricReader(
```
I think this way it actually sets it up as all cumulative, is that possible?

pichlermarc replied:
Yes, this was an oversight. Thought it was still working the way it did before, but it does not.
Added temporality preference in [2d01775](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/2d01775f272252daa002b671d1a760238bbcfc8c).

pirgeo commented on example/basic_example.py:
```diff
@@ -104,15 +106,16 @@ def parse_arguments():
     logger.debug("setting up global OpenTelemetry configuration.")
-    _metrics.set_meter_provider(MeterProvider(
+    metrics.set_meter_provider(MeterProvider(
         metric_readers=[PeriodicExportingMetricReader(
```
It seems to me that all of the Temporalities are `CUMULATIVE` if set up this way.

pichlermarc replied:
Yes, this was an oversight. Thought it was still working the way it did before, but it does not.
Added temporality preference in [2d01775](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/2d01775f272252daa002b671d1a760238bbcfc8c).

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
+                        try:
+                            string_buffer.append(
+                                self._serializer.serialize(dt_metric))
+                            string_buffer.append("\n")
```
Ah right, done in f3130f3.Thanks :slightly_smiling_face:

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
+                            self.__logger.warning(
+                                "Failed to serialize metric. Skipping: %s", ex)
+
+                        if len(string_buffer) / 2 >= chunk_size:
```
Also done in f3130f3. Thanks :slightly_smiling_face:

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
+                            try:
+                                self._send_lines(string_buffer)
+                                string_buffer = []
+                            except Exception as ex:
```
Right, `HTTPError` was not raised before this PR and still is raised with this code. I removed the `raises` comment from the doc.

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
+        except Exception as ex:
+            self.__logger.warning(
+                "Failed to export metrics: %s", ex)
+            return MetricExportResult.FAILURE
```
Removed it in f3130f3. It was annoying me as well but I wanted the keep the spirit of the original. They are now collected and sent after the fact.

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
+                Raises
+                ------
+                HTTPError
+                    If one occurred
```
removed in [f3130f3](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/f3130f3d6a6b3199a96d3e37b9498d58562ebbdb)

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
+
+                Parameters
+                ----------
+                timeout_millis
```
not used, removed in [f00198b](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/f00198b45eff167b6cd821ffa337938b3496fda2).

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
-    ) -> MetricExportResult:
+    def export(self,
+               metrics_data: MetricsData,
+               timeout_millis: float = 10_000,
```
not used, removed in [f00198b](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/f00198b45eff167b6cd821ffa337938b3496fda2).

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
+                    resp.content.decode("utf-8")))
+
+    def _sum_to_dynatrace_metric(self, metric: Metric, point: NumberDataPoint):
+        if isinstance(point.value, float):
```
Distinction for monotonic and non-monotonic sums added here: [2d01775](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/2d01775f272252daa002b671d1a760238bbcfc8c)

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
     def preferred_temporality(self) -> AggregationTemporality:
         return AggregationTemporality.DELTA
```
True, we don't need it anymore. This has changed. Removed in [2d01775](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/2d01775f272252daa002b671d1a760238bbcfc8c).

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -14,37 +14,39 @@
+    HistogramDataPoint
+)
 
 VERSION = "0.3.0b0"
```
Changed in [03f9577](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/03f9577cd87fd524977825e83dabae70eb2d2fb2)

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
+                Returns
+                -------
+                MetricExportResult
+                    Indicates SUCCESS or FAILURE
```
Addressed in [47d7ff9](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/47d7ff9ef78d21da4140434e069c30afa08415b0), slightly reworded though. PTAL. :slightly_smiling_face:

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -14,37 +14,39 @@
 
 
-def _get_histogram_max(histogram: Histogram):
+def _get_histogram_max(histogram: HistogramDataPoint):
```
I am aware of this, but I think extracting a method would hurt readability, as the current code needs to interrupt the execution flow. 

However, since `HistogramDataPoint` now has a `count` field, I used that to reduce the duplicated code to 13 lines (6 comments, one empty) in [e1795a2](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/e1795a2547b0a21ee1e0d219a0dc73f3f6316bbf) :slightly_smiling_face:

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -170,117 +176,135 @@ def __init__(
+                metric.name,
+                point.value,
+                dict(point.attributes),
+                int(point.time_unix_nano / 1000000))
```
I tried it out, and for every instrument adding/recording something else than an `int` or `float` has caused an exception long before it reaches the exporter.

pirgeo replied:
Okay!

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -239,19 +222,40 @@ def export(self,
+        chunk_size = DynatraceMetricsApiConstants.payload_lines_limit()
 
-    def _sum_to_dynatrace_metric(self, metric: Metric, point: NumberDataPoint):
+        for index in range(0, len(metric_lines), chunk_size):
```
💡 thats pretty cool

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -239,19 +222,40 @@ def export(self,
+        self.__logger.warning("Failed to create Dynatrace metric: "
+                              "exporter received %s '%s' with "
+                              "AggregationTemporality.%s, but only "
+                              "AggregationTemporality.%s is supported.",
```
```suggestion
                              "AggregationTemporality.%s is currently supported.",
```
Maybe?

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -239,19 +222,40 @@ def export(self,
+                              metric.data.aggregation_temporality.name,
+                              supported_temporality.name)
+
+    def _monotonic_to_dynatrace_metric(self, metric: Metric,
```
```suggestion
    def _monotonic_sum_to_dynatrace_metric(self, metric: Metric,
```

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -265,8 +269,8 @@ def _sum_to_dynatrace_metric(self, metric: Metric, point: NumberDataPoint):
 
-    def _gauge_to_dynatrace_metric(self, metric: Metric,
-                                   point: NumberDataPoint):
+    def _non_monotonic_to_dynatrace_metric(self, metric: Metric,
```
```suggestion
    def _non_monotonic_sum_to_dynatrace_metric(self, metric: Metric,
```

pichlermarc replied:
Here, sum was left out intentionally, as we also use this for gauges.

pirgeo replied:
Maybe we should just call it `_to_dynatrace_gauge` then?

pichlermarc replied:
Naming. :man_facepalming:  
[7e63c16](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/7e63c16965260103da8f48a1ad59c0afca470eb8)

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -282,6 +286,13 @@ def _gauge_to_dynatrace_metric(self, metric: Metric,
+        if metric.data.aggregation_temporality != AggregationTemporality.DELTA:
+            self._log_temporality_mismatch("Histogram",
+                                           metric,
+                                           AggregationTemporality.DELTA)
```
Nit: here you don't specify `supported_temporality` but below you do.

pichlermarc replied:
Fixed in [dabe990](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/dabe9905bcfdc77cced8aba647032ef43af50f66). :slightly_smiling_face:

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -291,14 +302,35 @@ def _histogram_to_dynatrace_metric(self, metric: Metric,
                 return self._histogram_to_dynatrace_metric(metric, point)
             if isinstance(metric.data, Gauge):
-                return self._gauge_to_dynatrace_metric(metric, point)
+                # allow any temporality.
```
```suggestion
                # gauge does not support or require temporality.
```

pichlermarc replied:
changed in [dabe990](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/dabe9905bcfdc77cced8aba647032ef43af50f66). Thanks!

pirgeo commented on test/test_exporter.py:
```diff
@@ -316,9 +340,38 @@ def test_sum_delta(self, mock_post):
+        mock_post.assert_not_called()
+
+    @patch.object(requests.Session, 'post')
+    def test_sum_cumulative_non_monotonic_exported_as_gauge(self, mock_post):
```
```suggestion
    def test_non_monotonic_cumulative_sum_exported_as_gauge(self, mock_post):
```
Its easier to read I think

pirgeo commented on test/test_exporter.py:
```diff
@@ -316,9 +340,38 @@ def test_sum_delta(self, mock_post):
-    def test_sum_delta_non_monotonic(self, mock_post):
-        metrics_data = self._metrics_data_from_data(
-            [self._create_sum(250, monotonic=False)])
+    def test_sum_delta_non_monotonic_is_dropped(self, mock_post):
```
```suggestion
    def test_non_monotonic_delta_sum_is_dropped(self, mock_post):
```

pirgeo commented on test/test_exporter.py:
```diff
@@ -301,7 +325,7 @@ def test_entire_batch_fail(self, mock_const, mock_post):
 
     @patch.object(requests.Session, 'post')
-    def test_sum_delta(self, mock_post):
+    def test_sum_delta_monotonic_exported_as_counter(self, mock_post):
```
```suggestion
    def test_monotonic_delta_sum_exported_as_counter(self, mock_post):
```

pirgeo commented on test/test_exporter.py:
```diff
@@ -316,9 +340,38 @@ def test_sum_delta(self, mock_post):
+        mock_post.assert_not_called()
+
+    @patch.object(requests.Session, 'post')
+    def test_sum_cumulative_monotonic_is_dropped(self, mock_post):
```
```suggestion
    def test_monotonic_cumulative_sum_is_dropped(self, mock_post):
```

pirgeo reviewed (commented):
Looking good. Waiting for the changes we discussed offline, but I think this is looking great already!

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -52,6 +53,27 @@
+        prefix: Optional[str] = None,
+        default_dimensions: Optional[Mapping[str, str]] = None,
+        export_dynatrace_metadata: Optional[bool] = False,
+        export_interval_millis: Optional[float] = None
```
Would it make sense to set the defaults here instead of further down in the code? This way, users of this method can just hover over the method and see what they will get if they don't specify anything.

pirgeo commented on test/test_exporter.py:
```diff
@@ -521,6 +523,63 @@ def test_view(self, mock_post):
+                    prefix="otel.python.test",
+                    default_dimensions={"defaultKey": "defaultValue"}
+                ),
+                                      PeriodicExportingMetricReader)
```
```suggestion
                    PeriodicExportingMetricReader)
```
I think?

pichlermarc replied:
fixed in [36cb2d6](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/36cb2d69d842f799aec8d23cf826184c545a7557).

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -37,12 +37,13 @@
 VERSION = "0.3.0-rc1"
 
-DYNATRACE_TEMPORALITY_PREFERENCE = {
+_DYNATRACE_TEMPORALITY_PREFERENCE = {
```
Yes it is. I also did not see any.

pichlermarc replied:
> Looks good to me! I have only one suggestion: The methods order inside `__init__.py` are a bit hard to follow/read. It would be nicer I think if they were positioned in the order they are called. E.g.
> 
>     1. `export`
> 
>     2. `_to_dynatrace_metric`
>        2.1 `_monotonic_sum_to_dynatrace_metric`
>        ...
> 
>     3. `_histogram_to_dynatrace_metric`
>        ...
> 
> 
> 4 `_send_lines`
> 
> Then the "utilities" one can be at the very bottom, like `_log_temporality_mismatch`
> 
> Also the ones on the very top `_get_histogram_max` can also be marked as static and placed at the bottom as well (or in a separated file, if you prefer).

I previously held off on moving things to another file as I thought it would blow up the diff too much. Since now two reviewers asked for it, I split this up into `exporter`, `histogram_utils` and `factory` in [82e7561](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/82e756162b8a90cbe1e254e2ce99afa57ad7cc2a). I also changed the ordering in [3d0c974](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/3d0c974f7d8496dcab142d516d92238442d13c4b).

joaopgrassi reviewed (approved):
Looks good to me! 
I have only one suggestion: The methods order inside `__init__.py` are a bit hard to follow/read. It would be nicer I think if they were positioned in the order they are called. E.g. 

1. `export`
2. `_to_dynatrace_metric`
2.1 `_monotonic_sum_to_dynatrace_metric`
...
3. `_histogram_to_dynatrace_metric`
...

4 `_send_lines`

Then the "utilities" one can be at the very bottom, like `_log_temporality_mismatch`

Also the ones on the very top `_get_histogram_max` can also be marked as static and placed at the bottom as well (or in a separated file, if you prefer).

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -166,121 +201,166 @@ def __init__(
+
+        for index in range(0, len(metric_lines), chunk_size):
+            metric_lines_chunk = metric_lines[index:index + chunk_size]
+            serialized_records = "\n".join(metric_lines_chunk) + "\n"
```
Good catch. The existing tests required it, MINT line protocol does not.

pichlermarc replied:
removed the newlines in [82e7561](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/82e756162b8a90cbe1e254e2ce99afa57ad7cc2a).

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -52,6 +53,27 @@
 }
 
 
+def configure_dynatrace_exporter(
```
I split it up in [82e7561](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/82e756162b8a90cbe1e254e2ce99afa57ad7cc2a). I previously held off on it because the diff gets huge with things like that.

pichlermarc commented on example/basic_example.py:
```diff
@@ -102,18 +101,19 @@ def parse_arguments():
-    # and the Dynatrace exporter exporting to args.endpoint with args.token
     logger.debug("setting up global OpenTelemetry configuration.")
+    # This call sets up the MeterProvider, with a
+    # PeriodicExportingMetricReader that exports every 5000 ms and the
```
fixed in [1bcd97e](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/1bcd97ebec651840d7f649345cba9f7a9d868cfd)

pichlermarc commented on example/basic_example.py:
```diff
@@ -102,18 +101,19 @@ def parse_arguments():
-    # This call sets up the MeterProvider, with a PeriodicExportingMetricReader that exports every 5000 ms
-    # and the Dynatrace exporter exporting to args.endpoint with args.token
     logger.debug("setting up global OpenTelemetry configuration.")
+    # This call sets up the MeterProvider, with a
```
fixed in [1bcd97e](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/1bcd97ebec651840d7f649345cba9f7a9d868cfd)

pichlermarc commented:
> Looks good to me! I have only one suggestion: The methods order inside `__init__.py` are a bit hard to follow/read. It would be nicer I think if they were positioned in the order they are called. E.g.
> 
>     1. `export`
> 
>     2. `_to_dynatrace_metric`
>        2.1 `_monotonic_sum_to_dynatrace_metric`
>        ...
> 
>     3. `_histogram_to_dynatrace_metric`
>        ...
> 
> 
> 4 `_send_lines`
> 
> Then the "utilities" one can be at the very bottom, like `_log_temporality_mismatch`
> 
> Also the ones on the very top `_get_histogram_max` can also be marked as static and placed at the bottom as well (or in a separated file, if you prefer).

I previously held off on moving things to another file as I thought it would blow up the diff too much. Since now two reviewers asked for it, I split this up into `exporter`, `histogram_utils` and `factory` in [82e7561](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/82e756162b8a90cbe1e254e2ce99afa57ad7cc2a). I also changed the ordering in [3d0c974](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/3d0c974f7d8496dcab142d516d92238442d13c4b).

pichlermarc commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -166,121 +201,166 @@ def __init__(
+            metric_type:
+            str, metric: Metric,
+            supported_temporality: AggregationTemporality):
+        self.__logger.warning("Failed to create Dynatrace metric: "
```
AFAIK this one and the collector exporter are the only ones that log on dropped metrics due to aggregation temporality. I would still keep this one, as it is more verbose as the one in the collector exporter.

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -61,6 +44,44 @@ def configure_dynatrace_exporter(
+        (default: no API token).
+    prefix: str, Optional
+        Will be prepended to each metric key, separated by a dot
+        (`<prefix>.<namespace>.<name>`).
```
```suggestion
        (`<prefix>.<name>`).
```
Otherwise people might be wondering where the `<namespace>.` is coming from

pichlermarc replied:
thx. removed in [2f5bcd7](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/2f5bcd7c53a60ca3de780552d59598a7071c66ee). also removed it from the readme where I copied it from.

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/_exporter.py:
```diff
@@ -0,0 +1,146 @@
+# Copyright 2020 Dynatrace LLC
```
```suggestion
# Copyright 2022 Dynatrace LLC
```
I think new files get the year in which they were first released as their Copyright year

pichlermarc replied:
fixed in [0ce3ad9](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/0ce3ad925520917615a5767d9205f81dda0762a5) :slightly_smiling_face:

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/__init__.py:
```diff
@@ -12,35 +12,18 @@
+
 VERSION = "0.3.0-rc1"
 
 _DYNATRACE_TEMPORALITY_PREFERENCE = {
```
Now there is no way for users to set this anymore, right? If that is the case, we can just keep it private.

pichlermarc replied:
Isn't it already private? :thinking:

pirgeo replied:
Yup! I just wanted to make sure that I understood correctly that this cannot be set anymore when using the DynatraceExporter.

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/_factory.py:
```diff
@@ -0,0 +1,154 @@
+# Copyright 2020 Dynatrace LLC
```
```suggestion
# Copyright 2022 Dynatrace LLC
```

pichlermarc replied:
fixed in [0ce3ad9](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/0ce3ad925520917615a5767d9205f81dda0762a5) :slightly_smiling_face:

pirgeo commented on src/dynatrace/opentelemetry/metrics/export/_histogram_utils.py:
```diff
@@ -0,0 +1,80 @@
+import math
```
Missing copyright!

joaopgrassi replied:
🤖

pirgeo reviewed (approved):
Great work!

pichlermarc commented on test/test_exporter.py:
```diff
@@ -44,7 +44,7 @@
     _DynatraceMetricsExporter,
     _DYNATRACE_TEMPORALITY_PREFERENCE,
-    configure_dynatrace_exporter
+    configure_dynatrace_export
```
good point. changed in [9011fff](https://github.com/dynatrace-oss/opentelemetry-metric-python/pull/27/commits/9011fff261563a6ac803c5ae7384852363a1e9b4). :slightly_smiling_face:
